Multiple definitions of sort_decays_list

When linking multiple model libraries to one executable, the following linking error occurs:

~~~.sh
/usr/bin/ld: /home/avoigt/research/FlexibleSUSY/models/SMEFTHiggsTopDown/libSMEFTHiggsTopDown.a(SMEFTHiggsTopDown_slha_io.o): in function `flexiblesusy::sort_decays_list(flexiblesusy::Decays_list const&)':
SMEFTHiggsTopDown_slha_io.cpp:(.text+0x8a80): multiple definition of `flexiblesusy::sort_decays_list(flexiblesusy::Decays_list const&)'; /home/avoigt/research/FlexibleSUSY/models/SM/libSM.a(SM_slha_io.o):SM_slha_io.cpp:(.text+0x8b80): first defined here
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [test/module.mk:1315: test/test_SMEFTHiggsTopDown.x] Error 1
~~~

The issue is that the function
~~~.cpp
std::vector<Decay> sort_decays_list(const Decays_list& decays_list) { ... }
~~~
is not local to the translation unit.

A possible solution would be to move this function into an anonymous namespace.

This issue can be reproduced as follows:
~~~.sh
git co feature-FlexiblEFTHiggs-3L
./createmodel --name=SM -f
./createmodel --name=SMEFTHiggsTopDown -f
./configure --with-models=SM,SMEFTHiggsTopDown --with-optional-modules=test
make -j2
make all-test -k
~~~

--- a/src/decays/decay.cpp
+++ b/src/decays/decay.cpp
@@ -115,6 +115,27 @@ const Decay& Decays_list::get_decay(
    return pos->second;
 }
 
+/**
+ * Sort decays of every particle according to their width
+ */
+std::vector<Decay> sort_decays_list(const Decays_list& decays_list) {
+   std::vector<Decay> decays_list_as_vector;
+   decays_list_as_vector.reserve(decays_list.size());
+   for (const auto& el : decays_list) {
+      decays_list_as_vector.push_back(el.second);
+   }
+
+   std::sort(
+      decays_list_as_vector.begin(),
+      decays_list_as_vector.end(),
+      [](const auto& d1, const auto& d2) {
+         return d1.get_width() > d2.get_width();
+      }
+   );
+
+   return decays_list_as_vector;
+}
+
 std::string strip_field_namespace(std::string const& s) {
    std::string result = s.substr(s.find_last_of(':')+1);
    if (s.find("bar") != std::string::npos) {

--- a/src/decays/decay.hpp
+++ b/src/decays/decay.hpp
@@ -95,6 +95,9 @@ private:
    double total_width{0.};
 };
 
+/// sort decays w.r.t. their width
+std::vector<Decay> sort_decays_list(const Decays_list&);
+
 std::string strip_field_namespace(std::string const&);
 
 template<typename FieldIn, typename FieldOut1, typename FieldOut2>

--- a/templates/slha_io.cpp.in
+++ b/templates/slha_io.cpp.in
@@ -56,32 +56,6 @@
 
 namespace flexiblesusy {
 
-namespace {
-
-/**
- * Sort decays of every particle according to their width
- *
- */
-std::vector<Decay> sort_decays_list(const Decays_list& decays_list) {
-   std::vector<Decay> decays_list_as_vector;
-   decays_list_as_vector.reserve(decays_list.size());
-   for (const auto& el : decays_list) {
-      decays_list_as_vector.push_back(el.second);
-   }
-
-   std::sort(
-      decays_list_as_vector.begin(),
-      decays_list_as_vector.end(),
-      [](const auto& d1, const auto& d2) {
-         return d1.get_width() > d2.get_width();
-      }
-   );
-
-   return decays_list_as_vector;
-}
-
-} // anonymous namespace
-
 @ModelName@_slha_io::@ModelName@_slha_io()
    : slha_io()
    , print_imaginary_parts_of_majorana_mixings(false)

--- a/templates/slha_io.cpp.in
+++ b/templates/slha_io.cpp.in
@@ -56,6 +56,32 @@
 
 namespace flexiblesusy {
 
+namespace {
+
+/**
+ * Sort decays of every particle according to their width
+ *
+ */
+std::vector<Decay> sort_decays_list(const Decays_list& decays_list) {
+   std::vector<Decay> decays_list_as_vector;
+   decays_list_as_vector.reserve(decays_list.size());
+   for (const auto& el : decays_list) {
+      decays_list_as_vector.push_back(el.second);
+   }
+
+   std::sort(
+      decays_list_as_vector.begin(),
+      decays_list_as_vector.end(),
+      [](const auto& d1, const auto& d2) {
+         return d1.get_width() > d2.get_width();
+      }
+   );
+
+   return decays_list_as_vector;
+}
+
+} // anonymous namespace
+
 @ModelName@_slha_io::@ModelName@_slha_io()
    : slha_io()
    , print_imaginary_parts_of_majorana_mixings(false)
@@ -407,28 +433,6 @@ void @ModelName@_slha_io::set_dcinfo(
           << FORMAT_SPINFO(9, SARAH_VERSION);
 
    slha_io.set_block(dcinfo);
-}
-
-/**
- * Sort decays of every particle according to their width
- *
- */
-std::vector<Decay> sort_decays_list(const Decays_list& decays_list) {
-   std::vector<Decay> decays_list_as_vector;
-   decays_list_as_vector.reserve(decays_list.size());
-   for (const auto& el : decays_list) {
-      decays_list_as_vector.push_back(el.second);
-   }
-
-   std::sort(
-      decays_list_as_vector.begin(),
-      decays_list_as_vector.end(),
-      [](const auto& d1, const auto& d2) {
-         return d1.get_width() > d2.get_width();
-      }
-   );
-
-   return decays_list_as_vector;
 }
 
 /**

